### ISSUE TITLE
Universal usage

### ISSUE THREAD

**Issue by dferber90** [css-modules/css-modules#97]
First of all, css modules are awesome! Thank you everyone! :)

I'm currently building a starter pack for web-apps. The goal is to have fast initial (server-side) renders and code splitting. I wrote a custom loader to support my use-case. I would like to verify my approach makes sense before publishing it. Maybe you'd even want it to be part of css-modules?

#### setup
I have got two webpack configurations. One builds for the client, the other one for the server.

```js
// client loader:
{ test: /\.css$/, loader: 'style!css?modules&localIdentName=[name]__[local]' },

// server loader
{ test: /\.css$/, loader: './myUniversalLoader.js!css?modules&localIdentName=[name]__[local]' },
```

On the client, I am using the style loader as usual. Nothing fancy happens. The css is contained in the JS bundles and code-split along with them.

The markup generated on the server should match the one the client generates later on. I wrote a [custom loader](https://github.com/dferber90/webapp/blob/760379fcc70ab451e5a239c04f888bb003a7ef22/modules/containers/mySecondLoader.js) which exports the `locals` information returned from `css?modules`. It is the counterpart of the `style` loader, which works on the server. It's job is to export the `locals`, but also make the underlying css available.

#### usage
Using `myUniversalLoader ` the markup from the server matches that from the client. However, because the client's CSS is contained in the JS bundle, a flash of unstyled text happens until the JS is loaded, parsed and executed. To prevent this, the JS could be moved into the head, but that would negate the effects of SSR (time-to-first-render). So, I decided to inline the initial CSS in the generated HTML page by matching the components in the router and extracting their style information [[details](https://github.com/dferber90/webapp/blob/760379fcc70ab451e5a239c04f888bb003a7ef22/modules/containers/HelloWorld.js#L18-L20)].

When the client bundle is loaded, the server-rendered inlined CSS styles are removed from the DOM and the client continues as usual (with style loader).

#### pros and cons
+ fast initial time-to-first-render
+ prevents flash of unstyled text

- CSS which is used by the components rendered initially has to be loaded twice (once inlined and once with the JS bundle)

More about the approach can be read [here](https://github.com/dferber90/webapp/blob/dd74328e57b4a1075505dbcbd1b5cf0ad5925372/knowhow/css-modules.md#css-modules-1).

#### questions
- Does something like `myUniversalLoader` already exist?
- Is there an easier way?
- Could we prevent loading the CSS of initial components twice?
- Should this be part of the css-modules organization?

**Comment by gajus** [comment]
@dferber90 Did you get anywhere further with this?

**Comment by dferber90** [comment]
@gajus Yeah, I released it as https://github.com/dferber90/fake-style-loader. But then `style-loader` caught up and so I deprecated it again.

The only additional value it provides is exporting the source of the used files as `source`. This can then be used for inlining in the HTML response.

A pretty rough example of this is in https://github.com/dferber90/webapp, as described in https://github.com/dferber90/webapp/blob/master/knowhow/css-modules.md#flash-of-unstyled-text.

I didn't take this to production anywhere so it's just a few unfinished experiments which one can build upon. Hope that helps :)
